modules/infra/datapath/control_input.c

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include "gr_control_input.h"
+
+#include <gr_eth_input.h>
+#include <gr_eth_output.h>
+#include <gr_graph.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+
+#include <rte_ether.h>
+#include <rte_graph_worker.h>
+#include <rte_malloc.h>
+#include <rte_version.h>
+
+enum {
+	UNKNOWN_CONTROL_INPUT_TYPE,
+	EDGE_COUNT,
+};
+
+struct gr_control_input_msg {
+	uint16_t type;
+	void *data;
+};
+
+static struct rte_ring *control_input_ring;
+static struct rte_mempool *control_input_pool;
+
+static rte_edge_t control_input_edges[1 << 8] = {UNKNOWN_CONTROL_INPUT_TYPE};
+
+void gr_control_input_add_handler(control_input_t type, const char *node_name) {
+	rte_edge_t edge = gr_node_attach_parent("control_input", node_name);
+	if (edge == RTE_EDGE_ID_INVALID)
+		ABORT("gr_node_attach_parent(control_input, lldp_output) failed");
+	if (control_input_edges[type] != UNKNOWN_CONTROL_INPUT_TYPE) {
+		ABORT("control_input: type=0x%04x -> already set to edge %u", type, edge);
+	}
+	control_input_edges[type] = edge;
+	LOG(DEBUG, "control_input: type=0x%04x -> edge %u", type, edge);
+}
+
+int control_input_post(control_input_t type, void *data) {
+	struct gr_control_input_msg msg;
+	int ret;
+
+	msg.type = type;
+	msg.data = data;
+
+	ret = rte_ring_enqueue_elem(control_input_ring, &msg, sizeof(msg));
+	if (ret < 0) {
+		goto err;
+	}
+	return 0;
+err:
+	return errno_set(-ret);
+}
+
+static uint16_t
+control_input_process(struct rte_graph *graph, struct rte_node *node, void **, uint16_t) {
+	struct gr_control_input_msg msg[RTE_GRAPH_BURST_SIZE];
+	struct rte_mbuf *mbuf;
+	uint16_t n;
+
+	n = rte_ring_dequeue_burst_elem(
+		control_input_ring,
+		msg,
+		sizeof(struct gr_control_input_msg),
+		RTE_GRAPH_BURST_SIZE,
+		NULL
+	);
+
+	for (unsigned i = 0; i < n; i++) {
+		mbuf = rte_pktmbuf_alloc(control_input_pool);
+		if (mbuf) {
+			control_input_mbuf_data(mbuf)->data = msg[i].data;
+			rte_node_enqueue_x1(graph, node, control_input_edges[msg[i].type], mbuf);
+		}
+	}
+
+	return n;
+}
+
+static void control_input_register(void) {
+	control_input_ring = rte_ring_create(
+		"control_input",
+		RTE_GRAPH_BURST_SIZE * 4,
+		SOCKET_ID_ANY,
+		RING_F_MP_RTS_ENQ | RING_F_MC_RTS_DEQ
+	);
+	if (control_input_ring == NULL)
+		ABORT("rte_ring_create(arp_output_request): %s", rte_strerror(rte_errno));
+
+	control_input_pool = rte_pktmbuf_pool_create(
+		"control_input",
+		RTE_GRAPH_BURST_SIZE * 4,
+		256, // cache_size
+		0, // priv_size
```

Comment:
Since commit https://github.com/rjarry/grout/commit/a12f55fb1cad5fe9620146fd2d50bb4b61a356bf you will need to specify the correct priv_size.

We need to move this into a single function to make sure we don't forget it.

Reply:
A helper for mempool creation is probably better.